### PR TITLE
API changes

### PR DESCRIPTION
**Pull request by nkoep**
This PR collects a few proposed API changes. Note that this is still a WIP. Here's the current rundown:

- Rename the `ConjugateGradient` solver to `ConjugateGradients`
- Rename the `SteepestDescent` solver to `GradientDescent`
- Make `tangent` a required method for classes deriving from `Manifold` to support the upcoming `check_gradient` utility function (see #105)
- Let the `Product` manifold class accept a variable number of arguments rather than a sequence, i.e., `Product(manifold1, manifold2)` instead of `Product([manifold1, manifold2])`
- Unify point and tangent vector representations of the `FixedRankEmbedded` manifold

Unfortunately, the last point uncovered a new (old) issue. The implementation of the `FixedRankEmbedded(m, n, k)` manifold in this PR is based on Manopt's representation, which means points are factored into triples `(u, s, v)` corresponding to a truncated SVD. This means that `u` and `v` have shapes `m x k` and `n x k`, respectively, and `s` is a `k x k` diagonal matrix such that a point `X` in the ambient space (of `m x n` matrices) corresponds to `X = u @ s @ v.T`.

The problem with this is that the autodiff backends won't know about the diagonal structure of `s` in this representation. Consider the low-rank matrix approximation example. For the cost function
```python
def cost(u, s, v):
    X = u @ s @ v.T
    return np.linalg.norm(X - A) ** 2
```
the autodiff backends generate gradient functions which produce the same output as the function 
```python
def egrad(u, s, v):
    X = u @ s @ v.T
    gu = 2 * (X - A) @ v @ s
    gs = 2 * u.T @ (X - A) @ v
    gv = 2 * (X - A).T @ u @ s
    return gu, gs, gv
```
Most notably, since `s` is a `k x k` matrix, the gradient of `cost` w.r.t. `s` has the same shape as `s`. However, even when the value passed in for `s` is a diagonal matrix, the off-diagonal elements in `gs` won't necessarily be zero; the diagonal elements agree with the true gradient though. In our earlier implementation this wasn't an issue as we stored the singular values as an array of shape `(k,)`.

The problem with the old approach, however, was that the difference in representation of points on the manifold and tangent vectors led to an incompatibility of the `FixedRankEmbedded` manifold with the `TrustRegions` solver. In order to compute Hessian-vector products, we need to compute inner products between gradients and tangent vectors (see https://en.wikipedia.org/wiki/Hessian_automatic_differentiation#Reverse_Hessian-vector_products and, e.g., our autograd backend implementation https://github.com/pymanopt/pymanopt/blob/master/pymanopt/autodiff/backends/_autograd.py#L43). The necessary dot products are therefore ill-defined due to mismatching shapes of the numpy arrays.

In order to resolve both issues, it seems necessary to go back to our earlier point representation to implicitly enforce the diagonality of gradients w.r.t. `s`, and adjust the representation of tangent vectors.

### CODE REVIEW

**Comment by kellertuer** [comment]
May I ask why you plan to rename `SteepestDescent` to `GradientDescent`? I would oppose that its the argument that a “gradient based descent” might use the gradient to find a descent direction that is not necessarily the steepest (for what reason it might be), so I am in favour of a steepest descent as a name.

**Comment by nkoep** [comment]
> May I ask why you plan to rename `SteepestDescent` to `GradientDescent`? I would oppose that its the argument that a “gradient based descent” might use the gradient to find a descent direction that is not necessarily the steepest (for what reason it might be), so I am in favour of a steepest descent as a name.

To my knowledge, the two names are synonymous, with the difference that gradient descent is the much more commonly used term, no doubt due to the prevalence of its variants in ML these days. The proposed change aims to clarify that the solver really doesn't do anything more complicated than "moving in the (negative) direction of the gradient" (with the implied changes to make things work on manifolds).

**Comment by kellertuer** [comment]
Yes, they are quite similar, you're right (also about the necessary modifications on manifolds of course). Well, I think I still prefer steepest descent :)

**Comment by sylvchev** [comment]
> Make tangent a required method for classes deriving from Manifold to support the upcoming check_gradient utility function (see #105)

The proposed implemention of `check_gradient` work also with manifold classes that do not implement a tangent method. So you could cross this point out from your list ^^

As a side remark, it could be nice to update the examples on www.pymanopt.org page, in the sections _Setting up the cost function using XXX backend_ with the new decorators.

**Comment by nkoep** [comment]
> > Make tangent a required method for classes deriving from Manifold to support the upcoming check_gradient utility function (see #105)
> 
> The proposed implementation of `check_gradient` also works with manifold classes that do not implement a tangent method. So you could cross this point out from your list ^^

Oh, good point re the `tangent` method. Somehow I thought we raised an exception if it's not available. In that case I'm on the fence about what to do with the method. It's only really used in `FixedRankEmbedded`, and only to project a tuple of random matrices to the tangent space, which might as well be done via the `proj` method. I may revert the change then since it really adds duplicated code to each manifold class.

> As a side remark, it could be nice to update the examples on [www.pymanopt.org](http://www.pymanopt.org) page, in the sections _Setting up the cost function using XXX backend_ with the new decorators.

The next big point on my personal TODO list is actually a major documentation overhaul. This also includes moving the examples from the landing page to the sphinx documentation and updating them in the process.

**Review by sylvchev — APPROVED** [review]

File: pymanopt/manifolds/fixed_rank.py [hunk 1]
```diff
@@ -91,8 +110,8 @@ def __init__(self, m, n, k):
         self._stiefel_m = Stiefel(m, k)
         self._stiefel_n = Stiefel(n, k)
 
-        name = ("Manifold of {m}-by-{n} matrices with rank {k} and embedded "
-                "geometry".format(m=m, n=n, k=k))
+        name = ("Manifold of ({m} x {n}) matrices of rank {k}".format(
+            m=m, n=n, k=k))
```
[the remark this answers is not in the document]
Comment (nkoep) [reply pymanopt/manifolds/fixed_rank.py:114]: Already on the todo list, along with normalizing code formatting across the code base via black + isort. I plan to do this in a separate PR after this one though given the amount of changes.

File: pymanopt/solvers/conjugate_gradients.py [hunk 2]
```diff
@@ -14,7 +14,7 @@
     "FletcherReeves PolakRibiere HestenesStiefel HagerZhang".split())
 
 
-class ConjugateGradient(Solver):
+class ConjugateGradients(Solver):
```
[the remark this answers is not in the document]
Comment (nkoep) [reply pymanopt/solvers/conjugate_gradients.py:17]: That's true and it's been brought up before so I'll likely revert this change again, even though I still think `ConjugateGradients` is a more apt/common name.

File: setup.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def parse_requirements_file(filename):
             "Programming Language :: Python :: 3",
             "Programming Language :: Python :: 3.5",
             "Programming Language :: Python :: 3.6",
```
[the remark this answers is not in the document]
Comment (nkoep) [reply setup.py:62]: Good catch! We are still intending to support python 3.6, but 3.5 we already dropped. I'll actually add this in a new PR before we release 0.2.6.

**Comment by nkoep** [comment]
Since I have changed my mind on most of the suggested changes of this PR I'm closing this after all. Some of the changes are now in #160. The changes to the `FixedRankEmbedded` manifold would have had to be reverted again anyway due to the autodiff issues outlined above. This means we will stick with the original representation for now which makes the class still incompatible with second-order methods.

**Pull request closed without merging**